test/test_optim.py

```diff
@@ -2216,9 +2218,90 @@ def test_non_empty_state(self, device, dtype, optim_info):
             for state in optim.state.values():
                 self.assertGreater(len(state), 0)
 
+    @optims(optim_db, dtypes=[torch.float64])
+    def test_optimizers_differentiable_wrt_params(
+        self, device: str, dtype: torch.dtype, optim_info: OptimizerInfo
+    ):
+        def _diff_fn(
+            p: torch.Tensor,
+            grad: torch.Tensor,
+            opt_differentiable_state: dict[str, Any],
+            opt_class: type[Optimizer],
+            kwargs: dict[str, Any],
+            *ignored,
+        ) -> tuple[torch.Tensor, ...]:
+            # Ignored is the list of values in `opt_differentiable_state`, we do this
+            # for `gradcheck` to correctly track the state tensors as function inputs
+            # because otherwise it can't unpack the values in the `opt_differentiable_state`
+            # dict
+            p = p.clone()
+            p.grad = grad
+            opt_differentiable_state = {
+                k: v.clone() if isinstance(v, torch.Tensor) else v
+                for k, v in opt_differentiable_state.items()
+            }
+            opt = opt_class([p], **kwargs)
+            opt.state[p].update(opt_differentiable_state)
+            opt.step()
+            return (p,) + tuple(
+                v
+                for v in opt.state[p].values()
+                if isinstance(v, torch.Tensor) and v.requires_grad
+            )
 
-instantiate_device_type_tests(TestOptimRenewed, globals(), allow_mps=True)
+        def _get_optimizer_state(
+            optim_cls: type[Optimizer],
+            params: torch.Tensor,
+            optim_kwargs: dict[str, Any],
+        ) -> dict[str, Any]:
+            params.grad = torch.rand_like(params)
+
+            optim = optim_cls([params], **optim_kwargs)
+            optim.step()
+
+            states = optim.state_dict()["state"]
+            # Accesses state of the first (and only) parameter. When using SGD w/ no momentum, there is no state
+            state = states[0] if 0 in states else states
+            return state
+
+        optim_cls = optim_info.optim_cls
+        if "differentiable" not in optim_info.supported_impls:
+            self.skipTest(f"Differentiable {optim_cls.__name__} not supported")
+
+        for optim_input in optim_info.optim_inputs_func(device=device):
+            mock_params = torch.rand(10, requires_grad=True, device=device, dtype=dtype)
+            mock_optim_kwargs = optim_input.kwargs.copy()
```

Comment:
what's the diff between copy and deepcopy? / why is the copy needed at all? Calling optim constructors shouldn't change the kwargs, do they?